_source/_posts/2021-08-04-microprofile-content-negotaion.adoc

```diff
@@ -0,0 +1,751 @@
+---
+layout: blog_post
+title: "Content Negotiation with a Java MicroProfile Application"
+author: brian-demers
+by: advocate
+communities: [java]
+description: "Tutorial: API clients have different capabilities and needs! Learn how to make your server select the correct languages, media types, and compression!"
+tags: [content-negotiation, microservices, java, quarkus, jaxrs]
+tweets:
+- "Languages and media-types and compression, Oh My‼️ Return the correct bits from your JAX-RS services‼️"
+- "Do does your #Java-based REST API need to support multiple languages? Multiple response types? How about compression?  Learn how in this post 🧑‍🏫"
+- "Learn how to support multiple response types and languages in your Java REST API 💥"
+image: blog/microprofile-content-negotaion/social.png
+type: awareness
+---
+
+Content negotiation allows for an HTTP server to respond to different types of clients. Many modern clients expect a JSON response, but there may be a need to format responses differently, maybe XML for older clients or a binary format for newer ones. Content negotiation is the mechanism used to solve that problem and others, such as dealing with multiple languages and even compressing HTTP requests.
+
+In this post, I'll walk through building a simple Java MicroProfile application and explain how content negotiation works.
+
+== Prerequisites
+- https://adoptopenjdk.net/[Java 11+]
+- https://maven.apache.org/download.cgi[Apache Maven]
+- https://httpie.io/docs#installation[HTTPie]
+
+If you want to skip ahead, the source for this example is on https://github.com/oktadev/quarkus-content-negotiation-example[GitHub].
+
+== What is MicroProfile?
+
+Around 2016, the Java EE specifications had started to stagnate and suffered from long release cycles; as a result, newer web services APIs were falling behind. Created to spur innovation, the new https://microprofile.io/[MicroProfile] project addressed the need to manage the existing Java JAX-RS, CDI, and JSON-P APIs, along with other new APIs for the evolving micro-services world.
+
+Fast forward to today, Java EE has moved to https://jakarta.ee/[Jakarta EE] under the https://www.eclipse.org/org/foundation/[Eclipse Foundation] and many MicroProfile projects have matured into Jakarta EE projects.
+
+TIP: Some projects change root packages as they move between projects. Anything starting with `javax.\*` may have moved to `jakarta.*` or `org.eclipse.*`.
+
+== What is Content Negotiation?
+
+In server-driven content negotiation, the client makes a request to a server with instructions on the type of response it can handle. When possible, the server responds with the appropriate  format or returns a `406` or `415` status code.
```

Comment:
```suggestion
In server-driven content negotiation, the client makes a request to a server with instructions on the type of response it can handle. When possible, the server responds with the appropriate format or returns a `406` or `415` status code.
```